File: flutter/test/scripts_test.dart
import 'package:busha_pay/src/scripts.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('document-start scripts', () {
    test('bridge shim defines window.BushaPayBridge', () {
      expect(bushaPayBridgeShim, contains('window.BushaPayBridge'));
      expect(bushaPayBridgeShim, contains('flutter_inappwebview'));
    });

    test('message listener routes statuses to bridge payload types', () {
      expect(messageListenerScript, contains("status === 'INITIALIZED'"));
      expect(messageListenerScript, contains("status === 'CANCELLED'"));
      expect(messageListenerScript, contains("status === 'COMPLETED'"));
      expect(messageListenerScript, contains("type: 'ready'"));
      expect(messageListenerScript, contains("type: 'close'"));
      expect(messageListenerScript, contains("type: 'success'"));
    });

    test('deep-link enabler removes navigator.getInstalledRelatedApps', () {
      expect(deepLinkEnablerScript, contains('delete navigator.getInstalledRelatedApps'));
      expect(deepLinkEnablerScript, contains('configurable: true'));
    });
  });

  group('autoSelectScript', () {
    test('JSON-encodes the prefix as a JS string literal', () {
      final js = autoSelectScript('Busha');
      expect(js, contains('var target = "Busha"'));
    });

    test('escapes embedded quotes safely', () {
      final js = autoSelectScript('She said "hi"');
      expect(js, contains(r'var target = "She said \"hi\""'));
    });

    test('embeds the auto-select guard and observer', () {
      final js = autoSelectScript('X');
      expect(js, contains('__bushaPayAutoSelectDone'));
      expect(js, contains('MutationObserver'));
      expect(js, contains("[role=\"button\"]"));
    });

    test('clears the observer after a 10 second timeout', () {
      final js = autoSelectScript('X');
      expect(js, contains('setTimeout'));
      expect(js, contains('10000'));
    });
  });
}

File: android/pay-android/src/test/kotlin/co/busha/pay/internal/ScriptsTest.kt
package co.busha.pay.internal

import org.junit.Assert.assertTrue
import org.junit.Test

class ScriptsTest {

    @Test
    fun documentStartScripts_defineTheBridgeShim() {
        assertTrue(DOCUMENT_START_SCRIPTS.contains("window.BushaPayBridge"))
        assertTrue(DOCUMENT_START_SCRIPTS.contains("window.BushaPayAndroid"))
        assertTrue(DOCUMENT_START_SCRIPTS.contains(".postMessage"))
    }

    @Test
    fun documentStartScripts_routeCheckoutStatuses() {
        assertTrue(DOCUMENT_START_SCRIPTS.contains("status === 'INITIALIZED'"))
        assertTrue(DOCUMENT_START_SCRIPTS.contains("status === 'CANCELLED'"))
        assertTrue(DOCUMENT_START_SCRIPTS.contains("status === 'COMPLETED'"))
        assertTrue(DOCUMENT_START_SCRIPTS.contains("type: 'ready'"))
        assertTrue(DOCUMENT_START_SCRIPTS.contains("type: 'close'"))
        assertTrue(DOCUMENT_START_SCRIPTS.contains("type: 'success'"))
    }

    @Test
    fun documentStartScripts_removeGetInstalledRelatedApps() {
        assertTrue(DOCUMENT_START_SCRIPTS.contains("delete navigator.getInstalledRelatedApps"))
    }

    @Test
    fun documentStartScripts_defineInitCheckout() {
        assertTrue(DOCUMENT_START_SCRIPTS.contains("window.initCheckout"))
        assertTrue(DOCUMENT_START_SCRIPTS.contains("__bushaPayInitCheckoutDefined"))
    }

    @Test
    fun initCheckoutScript_wrapsPayloadAndReturnsTrue() {
        val js = initCheckoutScript("""{"a":1}""")
        assertTrue(js.contains("""initCheckout({"a":1})"""))
        assertTrue(js.trim().endsWith("true;"))
    }

    @Test
    fun autoSelectScript_jsonEncodesPrefix() {
        assertTrue(autoSelectScript("Busha").contains("var target = \"Busha\""))
    }

    @Test
    fun autoSelectScript_escapesEmbeddedQuotes() {
        assertTrue(autoSelectScript("She said \"hi\"").contains("\\\"hi\\\""))
    }

    @Test
    fun autoSelectScript_embedsGuardObserverAndTimeout() {
        val js = autoSelectScript("X")
        assertTrue(js.contains("__bushaPayAutoSelectDone"))
        assertTrue(js.contains("MutationObserver"))
        assertTrue(js.contains("10000"))
        assertTrue(js.trim().endsWith("true;"))
    }
}
